doc/engine.md

```diff
@@ -0,0 +1,215 @@
+on a wide range of devices.
+
+Visual features that do not work on all supported devices may be regarded
+as low priority. (This does not apply to pure optimizations.)
+
+### 3.1. Client visual settings
+
+While the client may toggle various visual features for the sake
+of personal preference and performance, games should not be
+forced to utilize or anticipate those that significantly
+change the way content is being rendered.
+
+Examples of visually neutral settings:
+* Screen resolution
+* Maximum FPS
+* Vertical sync
+* Anti-aliasing
+* Texture filtering
+* Stereoscopy
+
+The above settings affect the image only in minor ways,
+and are safe to apply to just about any content.
+
+
+Examples of non-neutral features that should be API gated:
+* Direct lighting
+* Realtime shadows
+* Bump-mapping
+* FOV changes
+
+Above features affect visuals or gameplay in significant ways,
+requiring some degree of gating in the scripting API.
+It's best if the language in the config descriptions is carefully
+chosen to use words such as "allow" rather than "enable", to
+avoid confusion.
+
+It does not suffice to mark a feature as "experimental" or have an explicit
+"force" setting - assume that the user will enable everything you put
+in front of them, even to their detriment. Don't expect every user to know
+what the options actually do.
+
+A third class of visual settings would be speedhacks, such as:
+* Opaque water
+* Opaque leaves
+* No waving
+
+With speedhacks, there is a clear understanding that they are sacrificing
+quality for performance, and possibly breaking things in the process.
+In future GUI redesigns they should probably be labeled more clearly as such.
+
+Finally, texture packs are out of scope for these considerations,
+as they are a deliberate mechanism to locally customize
+the appearance of content.
+
+## 4. Upgrading
+
+Occasionally, it may be impossible to improve the engine without
+breaking compatibility in some ways. While it is never acceptable
+to break compat without providing some way to restore the original
+conditions, it may sometimes be okay to require some action from
+a mod or game author to maintain compatibility with subsequent
+engine versions. However, significant changes to existing APIs
+should only happen in a major release. Scripts should at least
+compile and run unmodified between minor releases.
+
+The standard of reasonable burden on a developer when upgrading
+to a new engine version is outlined in the following section.
+
+### 4.1. Reasonable upgrade burden
+
+The amount of work required from a game developer or modder
+to upgrade their scripts must be reasonable, and the old
+functionality must not be degraded in noticeable ways.
+Features must not consume any resources of the client
+or the server after being disabled by game code.
+
+Examples of reasonable burdens:
+* Utilizing a new API function in a few places
+* Utilizing a new object or node flag
```

Comment:
Just to clarify these points are supposed to mean "[Having to] utilize a new flag [to get old behavior back]", right? That didn't use to fly.

Reply:
> That didn't use to fly.

I do recall being made to do that myself... see the other, longer post. I'm totally fine with hardening these rules so that it doesn't happen again.